### PR TITLE
72 make chemprop multiclass classification model

### PR DESCRIPTION
**Pull request by JenniferHem**

### CODE REVIEW

File: test_extras/test_chemprop/test_chemprop_pipeline.py [hunk 1]
```diff
@@ -306,3 +335,44 @@ def test_prediction(self) -> None:
...
+        proba = classification_model.predict_proba(mols)
+        self.assertEqual(len(pred), len(test_data_df))
+        self.assertEqual(proba.shape[1], 3)
+        self.assertEqual(proba.shape[0], len(test_data_df))
+
+        model_copy = joblib_dump_load(classification_model)
+        pred_copy = model_copy.predict(mols)
+        proba_copy = model_copy.predict_proba(mols)
+
+        nan_indices = np.isnan(pred)
```
Comment (JochenSiegWork) [inline test_extras/test_chemprop/test_chemprop_pipeline.py:367]: `np.isnan` rather returns a mask and not indices. The name is not correct.

File: test_extras/test_chemprop/test_chemprop_pipeline.py [hunk 2]
```diff
@@ -306,3 +335,44 @@ def test_prediction(self) -> None:
...
+        self.assertListEqual(nan_indices.tolist(), np.isnan(pred_copy).tolist())
+        self.assertTrue(np.allclose(pred[~nan_indices], pred_copy[~nan_indices]))
+
+        self.assertEqual(proba.shape, proba_copy.shape)
```
Comment (JochenSiegWork) [inline test_extras/test_chemprop/test_chemprop_pipeline.py:371]: I think the `self.assertEqual(pred.shape, pred_copy.shape)` check is missing

**Review by JochenSiegWork — CHANGES_REQUESTED** [review]

File: molpipeline/estimators/chemprop/models.py [hunk 3]
```diff
@@ -348,3 +348,118 @@ def __init__(
...
+        **params
+            The parameters to set.
+
+        Returns
+        -------
+        Self
+            The model with the new parameters.
+        """
+        super().set_params(**params)
+        if not self._is_multiclass_classifier():
```
[the remark this answers is not in the document]
Comment (JenniferHem) [reply molpipeline/estimators/chemprop/models.py:411]: I basically oriented myself at the existing ChempropClassifier where the check also occurs. I guess we would need to check it here as well as set params can be called after instantiation?

File: molpipeline/estimators/chemprop/models.py [hunk 4]
```diff
@@ -348,3 +348,118 @@ def __init__(
...
+            raise ValueError(
+                "ChempropMulticlassClassifier should contain more than 2 classes."
```
[the remark this answers is not in the document]
Comment (JenniferHem) [reply molpipeline/estimators/chemprop/models.py:413]: true, added the test for the classes as well.

File: molpipeline/estimators/chemprop/models.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (JochenSiegWork) [reply molpipeline/estimators/chemprop/models.py:411]: Yes, exactly. We need to make the check twice. In the `__init__` and in the `set_params`.

File: molpipeline/estimators/chemprop/models.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (JochenSiegWork) [reply molpipeline/estimators/chemprop/models.py:413]: technically the exception's message is incomplete, since only `predictors` of type `MulticlassClassificationFFN` are valid. For example, you could adapt the message to: "The model's predictor or the number of classes are invalid. Use a multiclass predictor and more than 2 classes."

File: molpipeline/estimators/chemprop/models.py [hunk 5]
```diff
@@ -348,3 +348,118 @@ def __init__(
...
+        ValueError
+            if the classes found in y are not matching n_classes or if the class labels do not start from 0 to n_classes-1
+        """
+        unique_y = np.unique(y)
+        log = []
+        if self.n_classes != len(unique_y):
+            log.append(
+                f"Given number of classes in init (n_classes) does not match the number of unique classes (found {unique_y}) in the target data."
+            )
+        if sorted(unique_y) != list(range(self.n_classes)):
```
[the remark this answers is not in the document]
Comment (c-w-feldmann) [reply molpipeline/estimators/chemprop/models.py:460]: This is a limitation of torch (to be specific the loss function [`torch.nn.cross_entropy`](https://pytorch.org/docs/stable/generated/torch.nn.functional.cross_entropy.html), which works with indices).

**Review by c-w-feldmann — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 6]
```diff
@@ -3,4 +3,5 @@ __pycache__
 molpipeline.egg-info/
 lib/
 build/
+lightning_logs/
```
[the remark this answers is not in the document]
Comment (JochenSiegWork) [reply .gitignore:6]: I also thought about this while reading the PR. It will not hurt much but is only relevant for the extras and not the core library. I think I am fine with keeping it.

File: molpipeline/estimators/chemprop/models.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (JochenSiegWork) [reply molpipeline/estimators/chemprop/models.py:460]: thanks for clarifying

File: test_extras/test_chemprop/test_chemprop_pipeline.py (hunk 1 above)
Comment (JenniferHem) [reply test_extras/test_chemprop/test_chemprop_pipeline.py:372]: again @c-w-feldmann should we change this for Classification Pipeline as well?

File: molpipeline/estimators/chemprop/models.py [hunk 7]
```diff
@@ -348,3 +348,118 @@ def __init__(
...
+            The input data.
+        y : Sequence[int | float] | npt.NDArray[np.int_ | np.float64]
+            The target data.
+
+        Returns
+        -------
+        Self
+            The fitted model.
+        """
+        self._check_correct_input(y)
```
[the remark this answers is not in the document]
Comment (JenniferHem) [reply molpipeline/estimators/chemprop/models.py:437]: This is as Christian explained below, a strange thing of torch itself. Scikit just assigns everything by itself, while torch requires classes to start at 0. Since the error torch raises is an "Index out of bounds" instad a "check your labels" I thought we should do this here.

File: test_extras/test_chemprop/test_chemprop_pipeline.py (hunk 1 above)
Comment (c-w-feldmann) [reply test_extras/test_chemprop/test_chemprop_pipeline.py]: yep

File: molpipeline/estimators/chemprop/models.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (JenniferHemmerich) [reply molpipeline/estimators/chemprop/models.py]: @c-w-feldmann @JochenSiegWork please check again. I split the checks for the class and the validation into two parts.

File: molpipeline/estimators/chemprop/models.py [hunk 8]
```diff
@@ -348,3 +348,130 @@ def __init__(
...
+        ----------
+        **params
+            The parameters to set.
+
+        Returns
+        -------
+        Self
+            The model with the new parameters.
+        """
+        super().set_params(**params)
```
Comment (JochenSiegWork) [inline molpipeline/estimators/chemprop/models.py:411]: I think `self.n_classes` should also be extracted from `params` and set here if given, since `n_classes` is an attribute of this instance and not the parent?  If you intentionally not set `n_classes` here I would be interested in the reason for doing so.

File: test_extras/test_chemprop/test_chemprop_pipeline.py [hunk 9]
```diff
@@ -306,3 +340,49 @@ def test_prediction(self) -> None:
...
+        self.assertEqual(pred.shape, pred_copy.shape)
+        self.assertTrue(np.allclose(proba[~nan_mask], proba_copy[~nan_mask]))
+
+        with self.assertRaises(ValueError):
+            classification_model.fit(
+                mols,
+                test_data_df["Label"].add(1).to_numpy(),
+            )
+        with self.assertRaises(ValueError):
```
Comment (JochenSiegWork) [inline test_extras/test_chemprop/test_chemprop_pipeline.py:383]: What you also could test here is that an ValueError is raised when you give a correct number of `n_classes` but not a MulticlassClassifier type to the constructor.

File: test_extras/test_chemprop/test_chemprop_pipeline.py [hunk 10]
```diff
@@ -306,3 +340,49 @@ def test_prediction(self) -> None:
...
+            classification_model = get_multiclass_classification_pipeline(n_classes=2)
+            classification_model.fit(
+                mols,
+                test_data_df["Label"].to_numpy(),
```
Comment (JochenSiegWork) [inline test_extras/test_chemprop/test_chemprop_pipeline.py:387]: Two small tests for the `set_params` method should also be added. One test should use valid input and the other should test invalid input that triggers `self._is_valid_multiclass_classifier()` method.

**Review by JochenSiegWork — CHANGES_REQUESTED** [review]
already look very good. Just two small additions to test the correct behaviour of `set_params`

File: molpipeline/estimators/chemprop/models.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (JochenSiegWork) [reply molpipeline/estimators/chemprop/models.py]: looks good to me

File: test_extras/test_chemprop/chemprop_test_utils/constant_vars.py [hunk 11]
```diff
@@ -94,3 +96,142 @@
...
+    "model__predictor__hidden_dim": 300,
+    "model__predictor__input_dim": 300,
+    "model__predictor__n_layers": 1,
+    "model__predictor__n_tasks": 1,
+    "model__predictor__threshold": None,
+    "n_jobs": 1,
+}
+
+
+DEFAULT_MULTICLASS_CLASSIFICATION_PARAMS = {
```
Comment (c-w-feldmann) [inline test_extras/test_chemprop/chemprop_test_utils/constant_vars.py:165]: Should we define a list with the intersection of DEFAULT_MULTICLASS_CLASSIFICATION_PARAMS  and DEFAULT_BINARY_CLASSIFICATION_PARAMS, an then extend it to both cases? This would be less redundant.

File: test_extras/test_chemprop/test_models.py [hunk 12]
```diff
@@ -177,3 +192,60 @@ def test_get_params(self) -> None:
...
+
+    def test_get_params(self) -> None:
+        """Test the get_params and set_params methods."""
+        n_classes = 3
+        chemprop_model = ChempropMulticlassClassifier(lightning_trainer__accelerator="cpu", n_classes=n_classes)
+        param_dict = chemprop_model.get_params(deep=True)
+        expected_params = dict(DEFAULT_MULTICLASS_CLASSIFICATION_PARAMS)  # Shallow copy
+        expected_params["model__predictor__n_classes"] = n_classes
+        expected_params["n_classes"] = n_classes
+        self.assertSetEqual(set(param_dict.keys()), set(expected_params.keys()))
```
Comment (c-w-feldmann) [inline test_extras/test_chemprop/test_models.py:211]: use `compare_params` from chemprop_test_utils/compare_models.py, see example [here](https://github.com/basf/MolPipeline/blob/main/test_extras/test_chemprop/test_neural_fingerprint.py#L28)

**Review by c-w-feldmann — CHANGES_REQUESTED** [review]

File: test_extras/test_chemprop/test_chemprop_pipeline.py [hunk 13]
```diff
@@ -139,6 +140,40 @@ def get_classification_pipeline() -> Pipeline:
     return model_pipeline
 
 
+def get_multiclass_classification_pipeline(n_classes: int) -> Pipeline:
```
[the remark this answers is not in the document]
Comment (JenniferHemmerich) [reply test_extras/test_chemprop/test_chemprop_pipeline.py:143]: why not move everything at once in the separate issue?

File: test_extras/test_chemprop/test_models.py (hunk 12 above)
Comment (JenniferHem) [reply test_extras/test_chemprop/test_models.py:211]: also needs to be adapted in other tests

File: test_extras/test_chemprop/chemprop_test_utils/constant_vars.py (hunk 11 above)
Comment (JenniferHem) [reply test_extras/test_chemprop/chemprop_test_utils/constant_vars.py:165]: maybe in another PR and then also define those fpr regresstion etc?

File: test_extras/test_chemprop/test_models.py (hunk 12 above)
Comment (JenniferHem) [reply test_extras/test_chemprop/test_models.py:211]: This does not work as we compare dicts and not models.

File: test_extras/test_chemprop/chemprop_test_utils/constant_vars.py (hunk 11 above)
Comment (c-w-feldmann) [reply test_extras/test_chemprop/chemprop_test_utils/constant_vars.py:165]: See [issue](https://github.com/basf/MolPipeline/issues/80)

**Review by c-w-feldmann — APPROVED** [review]
LGTM

**Review by JochenSiegWork — CHANGES_REQUESTED** [review]

**Review by JochenSiegWork — APPROVED** [review]
